### ISSUE TITLE
Multiple dispatch 

### ISSUE THREAD

**Issue by ShalokShalom** [kalai-transpiler/kalai#68]
Hi there 

Is it save to assume, that features that are not explicitly mentioned in the [documentation](https://github.com/kalai-transpiler/kalai/blob/main/docs/Syntax.md) is inimplicitly supported?

Multiple dispatch, in this case?

**Comment by echeran** [comment]
It's good to not assume any features that are not listed in the documentation. I think that's what you were saying, and I would agree.

As far as multiple dispatch, if I'm understanding the question correctly, we don't support that. We have some notes in [`docs/Syntax.md`](https://github.com/kalai-transpiler/kalai/blob/main/docs/Syntax.md) under the section [Functions / Overloads / Arities](https://github.com/kalai-transpiler/kalai/blob/main/docs/Syntax.md#functions--overloads--arities). We only support a function name having single dispatch, which was a constraint encountered early on when supporting Rust. (The Rust constraint, which is like C's constraint, allows Rust compiled binaries to have stable ABI like C binaries.)  So that prevents us from having overloaded functions like in Java (multi-artiy fns in Clojure) or polymorphic dispatch (ex: Clojure multimethods).

Did that address the question you were asking?

**Comment by ShalokShalom** [comment]
Yes, thanks. Traits would be no solution?

**Comment by echeran** [comment]
Traits are interfaces that allow methods to have default implementations, AFAIU. So that's analogous to Java 8+ interfaces. And Clojure has protocols to provide an interface a la carte (no default impls). Clojure protocols and Rust traits are more flexible than Java interfaces because they allow developer to choose where/how they can declare that a named record/type/struct implements the protocol/interface/trait. In Java, unless you edit the definition of a class that implements an interface, you can't make that class officially implement that interface.

So Java is already a tad deficient compared to Rust & Clojure in functionality. But if we think about languages like C++, C, Python, Ruby 2.x, JS, etc. Python has duck typing, Ruby 3 apparently introduced interfaces, the best you can do in C++ is abstract classes, and you're out of luck in C or JS.

Long story short, although we haven't thought through this issue until you asked (thanks!), it feels like one of those features that unfortunately is outside the intersection set of what is supported in many popular programming languages of the day. So it would be hard to support. While Rust is usually the outlier in such questions, here, it is not such at all!

That being said, designing to interfaces is emphasized quite a lot in the Clojure world, and that design aesthetic serves us very well outside of Clojure contexts, too. If our work so far has taught us that persistent data structures are the most impactful missing feature in popular programming languages, then my personal opinion is that Clojure style protocols (interfaces) might be in 2nd place to that.

**Comment by ShalokShalom** [comment]
Yeah, I guess you need dynamic runtime features for multiple dispatch, thanks for the deep explanation
